hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/DatePartitionPathSelector.java

```diff
@@ -0,0 +1,217 @@
+  public static class Config {
+    public static final String DATE_PARTITION_DEPTH =
+        "hoodie.deltastreamer.source.input.date_partition.selector.date_partition_depth";
+    public static final String NUM_PREV_DAYS_TO_LIST =
+        "hoodie.deltastreamer.source.input.date_partition.selector.num_prev_days_to_list";
+    public static final String CURRENT_DATE =
+        "hoodie.deltastreamer.source.input.date_partition.selector.current_date";
+    public static final String PARTITIONS_LIST_PARALLELISM =
+        "hoodie.deltastreamer.source.input.date_partition.selector.partitions.list.parallelism";
+    public static final int DEFAULT_DATE_PARTITION_DEPTH = 0; // Implies no (date) partition
+    public static final int DEFAULT_NUM_DAYS_TO_LIST = 2;
+    public static final int DEFAULT_PARTITIONS_LIST_PARALLELISM = 20;
+  }
+
+  public DatePartitionPathSelector(TypedProperties props, Configuration hadoopConf) {
+    super(props, hadoopConf);
+    /*
+     * datePartitionDepth = 0 is same as basepath and there is no partition. In which case
+     * this path selector would be a no-op and lists all paths under the table basepath.
+     */
+    datePartitionDepth = props.getInteger(DATE_PARTITION_DEPTH, DEFAULT_DATE_PARTITION_DEPTH);
+    // If not specified the current date is assumed by default.
+    currentDate = LocalDate.parse(props.getString(Config.CURRENT_DATE, LocalDate.now().toString()));
+    numPrevDaysToList = props.getInteger(NUM_PREV_DAYS_TO_LIST, DEFAULT_NUM_DAYS_TO_LIST);
+    fromDate = currentDate.minusDays(numPrevDaysToList);
+    partitionsListParallelism =
+        props.getInteger(PARTITIONS_LIST_PARALLELISM, DEFAULT_PARTITIONS_LIST_PARALLELISM);
+  }
+
+  @Override
+  public Pair<Option<String>, String> getNextFilePathsAndMaxModificationTime(
+      JavaSparkContext sparkContext, Option<String> lastCheckpointStr, long sourceLimit) {
+    try {
+      // obtain all eligible files under root folder.
+      log.info(
+          "Root path => "
+              + props.getString(ROOT_INPUT_PATH_PROP)
+              + " source limit => "
+              + sourceLimit
+              + " depth of day partition => "
+              + datePartitionDepth
+              + " num prev days to list => "
+              + numPrevDaysToList
+              + " from current date => "
+              + currentDate);
+      long lastCheckpointTime = lastCheckpointStr.map(Long::parseLong).orElse(Long.MIN_VALUE);
+      HoodieSparkEngineContext context = new HoodieSparkEngineContext(sparkContext);
+      List<String> prunedPaths =
+          pruneDatePartitionPaths(context, fs, props.getString(ROOT_INPUT_PATH_PROP));
+      List<FileStatus> eligibleFiles = new ArrayList<>();
+      for (String path : prunedPaths) {
+        eligibleFiles.addAll(listEligibleFiles(fs, new Path(path), lastCheckpointTime));
+      }
+      // sort them by modification time.
+      eligibleFiles.sort(Comparator.comparingLong(FileStatus::getModificationTime));
```

Comment:
Once again, this sorting can happen in parallel right? and just collect it finally